server/proxy/pf_client.c

```diff
@@ -216,73 +241,211 @@ static BOOL pf_client_pre_connect(freerdp* instance)
 		return FALSE;
 	}
 
-	return TRUE;
+	return pf_modules_run_hook(pc->pdata->module, HOOK_TYPE_CLIENT_PRE_CONNECT, pc->pdata, pc);
 }
 
 static BOOL pf_client_receive_channel_data_hook(freerdp* instance, UINT16 channelId,
                                                 const BYTE* data, size_t size, UINT32 flags,
                                                 size_t totalSize)
 {
-	pClientContext* pc = (pClientContext*)instance->context;
-	pServerContext* ps = pc->pdata->ps;
-	proxyData* pdata = ps->pdata;
-	proxyConfig* config = pdata->config;
-	size_t i;
-
 	const char* channel_name = freerdp_channels_get_name_by_id(instance, channelId);
+	pClientContext* pc;
+	pServerContext* ps;
+	proxyData* pdata;
+	const proxyConfig* config;
+	int pass;
+
+	WINPR_ASSERT(instance);
+	WINPR_ASSERT(data || (size == 0));
+
+	pc = (pClientContext*)instance->context;
+	WINPR_ASSERT(pc);
+	WINPR_ASSERT(pc->pdata);
+
+	ps = pc->pdata->ps;
+	WINPR_ASSERT(ps);
 
-	for (i = 0; i < config->PassthroughCount; i++)
+	pdata = ps->pdata;
+	WINPR_ASSERT(pdata);
+
+	config = pdata->config;
+	WINPR_ASSERT(config);
+
+	pass = pf_utils_channel_is_passthrough(config, channel_name);
+
+	switch (pass)
 	{
-		if (strncmp(channel_name, config->Passthrough[i], CHANNEL_NAME_LEN) == 0)
+		case 0:
+			return TRUE; /* Silently drop */
+		case 1:
 		{
 			proxyChannelDataEventInfo ev;
-			UINT64 server_channel_id;
+			UINT16 server_channel_id;
 
 			ev.channel_id = channelId;
 			ev.channel_name = channel_name;
 			ev.data = data;
 			ev.data_len = size;
-
-			if (!pf_modules_run_filter(FILTER_TYPE_CLIENT_PASSTHROUGH_CHANNEL_DATA, pdata, &ev))
-				return FALSE;
-
-			server_channel_id =
-			    (UINT64)HashTable_GetItemValue(ps->vc_ids, (const void*)channel_name);
-			return ps->context.peer->SendChannelData(ps->context.peer, (UINT16)server_channel_id,
-			                                         data, size);
+			ev.flags = flags;
+			ev.total_size = totalSize;
+
+			if (!pf_modules_run_filter(pdata->module, FILTER_TYPE_CLIENT_PASSTHROUGH_CHANNEL_DATA,
+			                           pdata, &ev))
+				return TRUE; /* Silently drop */
+
+			/* Dynamic channels need special treatment
+			 *
+			 * We need to check every message with CHANNEL_FLAG_FIRST set if it
+			 * is a CREATE_REQUEST_PDU (0x01) and extract channelId and name
+			 * from it.
+			 *
+			 * To avoid issues with (misbehaving) clients assume all packets
+			 * that do not have at least a length of 1 byte and all incomplete
+			 * CREATE_REQUEST_PDU (0x01) packets as invalid.
+			 */
+			if ((flags & CHANNEL_FLAG_FIRST) &&
+			    (strncmp(channel_name, DRDYNVC_SVC_CHANNEL_NAME, CHANNEL_NAME_LEN + 1) == 0))
+			{
+				BYTE cmd;
+				if (size < 1)
+					return FALSE;
+
+				cmd = data[0] >> 4;
+				if (cmd == 0x01)
+				{
+					proxyChannelDataEventInfo dev;
+					size_t len, nameLen;
+					const char* name;
+					UINT32 dynChannelId;
+					BYTE cbId = data[0] & 0x03;
+					switch (cbId)
+					{
```

Comment:
Here I don't like the way wire data are parsed, should use `Stream_Read_XXX` functions instead (using a `Stream_Static_Init` on `data`).

Reply:
too much overhead here, we just look at the first byte

Reply:
Well that's not entirely true as the name is also read, and the channel id. `Stream_Static_Init` is really light, that would remove the reading of channel id by hand, and the name would be retrieved by a simple `Stream_Pointer()`. I can't remember but I think we also have functions that read a null terminated string from a stream...
Except if I missed something and this function is the proxy's bottleneck, I don't see why we would do everything "by hand" here and use regular Stream functions everywhere else.